### ISSUE TITLE
[Bug Report] Send message to multiple outputs using ModuleClient throws

### ISSUE THREAD

**Issue by Sdelausnay** [Azure/azure-iot-sdk-csharp#3201]
# Context

- **OS, version, SKU and CPU architecture used:**  Raspbian GNU/Linux 9 (stretch), armv7l
- **Application's .NET Target Framework :** net6.0
- **Device:** Raspberry PI3
- **SDK version used:** Microsoft.Azure.Devices.Client 1.41.3

## Description of the issue
I have a message in a custom module which I want to send to 2 different output depending on the message content. Sending the message to the first output is no issue, but when targeting the second output gives and exception of  `An item with the same key has already been added. Key: iothub-outputname` . I have check the code but I'm sure that I'm sending the message to a different output name. 

## Code sample exhibiting the issue
```csharp

var messageJson = JsonConvert.SerializeObject(mqttMsg, JsonSerializerSettings);

var message = new Message(Encoding.UTF8.GetBytes(messageJson))
{
  ContentEncoding = "UTF-8",
  ContentType = "application/JSON"
};

_logger.LogTrace("Send to main output: {0}", Endpoints.MqttBrokerOutputName);

await _moduleClient.SendEventAsync(Endpoints.MqttBrokerOutputName, message, stoppingToken);

if (mqttMessages.FirstOrDefault().Topic.Contains("boxstatus"))
{
  _logger.LogTrace("Send to Events output: {0}", Endpoints.EventsModuleOutputName);
  await _moduleClient.SendEventAsync(Endpoints.EventsModuleOutputName, message, stoppingToken);
}

_logger.LogTrace("Messages send to all endpoints.");                 
```

## Console log of the issue 

Line 153 refers to the `await _moduleClient.SendEventAsync(Endpoints.EventsModuleOutputName, message, stoppingToken);`

<details><summary>Logs</summary>
<p>

```
<7>2023-03-27 22:08:53.293 +00:00 MqttBroker.Services.MqttBrokerService[0] Send to main output: mqtt
<7>2023-03-27 22:08:53.303 +00:00 MqttBroker.Services.MqttBrokerService[0] Send to Events output: events_output
<3>2023-03-27 22:08:53.309 +00:00 MqttBroker.Services.MqttBrokerService[0] An unexpected error occurred - terminating MqttBroker module System.ArgumentException: An item with the same key has already been added. Key: iothub-outputname    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)    at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)    at Microsoft.Azure.Devices.Client.ReadOnlyDictionary45`2.System.Collections.Generic.IDictionary<TKey,TValue>.Add(TKey key, TValue value)    at Microsoft.Azure.Devices.Client.InternalClient.SendEventAsync(String outputName, Message message, CancellationToken cancellationToken)    at MqttBroker.Services.MqttBrokerService.ExecuteAsync(CancellationToken stoppingToken) in /app/MqttBroker/Services/MqttBrokerService.cs:line 153

```

</p>
</details>

**Comment by drwill-ms** [comment]
@Sdelausnay this appears to be because of how the SDK sets the system property, using a Dictionary<K, V>.Add(). This means that subsequent calls to send that message will try to add that same value to the dictionary, which fails with the exception you've observed.

The work around would be to use the `Message.CloneWithBody(in byte[] array)` method before sending to either route.

I'll look at changing the code to set the dictionary key using an index setter, which should overwrite the output name if it was previously set. Note, a problem with this is that if the message is being sent to two routes at the same time (the task isn't awaited) then the output name will change for the first send part way through, which could cause problems. For that reason, and other possible issues I might run into, it may be best to duplicate the message instead of trying to reuse it.

**Comment by drwill-ms** [comment]
I checked our v2 preview and this bug wouldn't repro there, as we don't use the `Add()` method.

**Comment by timstokman** [comment]
The same issue happens when you reuse message objects, encountered this with a keep-alive style message that's periodically sent:

```
const string channel = "keep-alive";
Message message = new(Encoding.UTF8.GetBytes("{ \"type\": \"keep-alive\" }"))
{
    MessageId = Guid.NewGuid().ToString(),
    ContentType = "application/json",
    ContentEncoding = "UTF-8",
};
while (true)
{
    message.MessageId = Guid.NewGuid().ToString();
    await client.SendEventAsync(channel, message, default);
}
```

**Comment by drwill-ms** [comment]
This fix as shipped in version [1.42.0](https://www.nuget.org/packages/Microsoft.Azure.Devices.Client/1.42.0). Release notes are [here](https://github.com/Azure/azure-iot-sdk-csharp/releases/tag/2023-05-01).
